client/asset/eth/eth.go

```diff
@@ -2016,6 +2015,73 @@ func (w *TokenWallet) EstimateRegistrationTxFee(feeRate uint64) uint64 {
 		return math.MaxUint64
 	}
 	return g.Transfer * feeRate
+}
+
+// EstimateSendTxFee returns a tx fee estimate for sending provided amount using
+// the provided feeRate.
+func (w *ETHWallet) EstimateSendTxFee(value, _ uint64, subtract bool) (fee uint64, err error) {
+	if subtract {
+		return 0, fmt.Errorf("wallet does not support checking network fee for withdrawal")
+	}
+	bal, err := w.Balance()
+	if err != nil {
+		return 0, nil
+	}
+	avail := bal.Available
+	if avail < value {
+		return 0, fmt.Errorf("not enough funds to send: have %d gwei need %d gwei", avail, value)
+	}
+
+	maxFeeRate, err := w.recommendedMaxFeeRate(w.ctx)
+	if err != nil {
+		return 0, fmt.Errorf("error getting max fee rate: %w", err)
+	}
+
+	maxFee := defaultSendGasLimit * dexeth.WeiToGwei(maxFeeRate)
+	if avail < value+maxFee {
+		return 0, fmt.Errorf("not enough funds to send: cannot cover value %d, max fee of %d gwei", value, maxFee)
+	}
+
+	return maxFee, nil
+}
+
+// EstimateSendTxFee returns a tx fee estimate for sending provided amount using
+// the provided feeRate.
+func (w *TokenWallet) EstimateSendTxFee(value, _ uint64, subtract bool) (fee uint64, err error) {
+	if subtract {
+		return 0, fmt.Errorf("wallet does not support checking network fee for withdrawal.")
+	}
+	bal, err := w.Balance()
+	if err != nil {
+		return 0, nil
+	}
+	avail := bal.Available
+	if avail < value {
+		return 0, fmt.Errorf("not enough tokens: have %d gwei need %d gwei", avail, value)
+	}
```

Comment:
it seems `avail < value` is comparing amount of tokens (e.g. USDC), while error message mentions gwei (which is kinda unrelated here)

Reply:
eth tokens are measured in `gwei`?

Reply:
Nope, Ethereum tokens (aka ERC20 tokens - functionality TokenWallet implements) are measured in their own units (e.g. USDC) while 1 wei is just a short for 1 * 10^(-18) ETH (and gwei is short for 1 * 10^(-9) ETH),

for example, USDC token has 6 decimal points of precision, thus its smallest unit is 0.000001 USDC and I don't think it has an equivalent short-cut name such as gwei

Reply:
I think this is handled by https://github.com/decred/dcrdex/blob/b7f667d97c3f59e6d5e4a0ad42fa7911a06fb87d/dex/networks/eth/tokens.go#L22-L28

Whenever we register the token, we could set that to 0-4 for USDC, for example, and it should all fall into place. Could of course have bugs. Testing this hard SOON. Really it depends on whether Max Total Supply in the token's smallest unit can overflow an int64 more so than the decimals they have listed. Should be ok to use 0 for USDC.

I think you are right that `gwei` doesn't cover everything... Not even sure if it makes sense for tokens with the same decimal places. Maybe just `units`?

Reply:
thank you for clarification, `units` look reasonable to me